src/main/java/org/opensearch/security/auth/BackendRegistry.java

```diff
@@ -387,14 +391,11 @@ public boolean authenticate(final SecurityRequestChannel request) {
 
         if (authenticated) {
             final User impersonatedUser = impersonate(request, authenticatedUser);
-            threadPool.getThreadContext()
-                .putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, impersonatedUser == null ? authenticatedUser : impersonatedUser);
-            auditLog.logSucceededLogin(
-                (impersonatedUser == null ? authenticatedUser : impersonatedUser).getName(),
-                false,
-                authenticatedUser.getName(),
-                request
-            );
+            final User effectiveUser = impersonatedUser == null ? authenticatedUser : impersonatedUser;
+            threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, effectiveUser);
+            UserSubject subject = new SecurityUser(threadPool, effectiveUser);
+            threadPool.getThreadContext().putPersistent(ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER, subject);
```

Comment:
I'm seeing the `putPersistent()` method for the first time. I have read in the API docs that the difference is that the value cannot be stashed. 

This kind of shows me that I do not really understand the sematic differences between `UserSubject` and `User`? Why is one persistent, but the other transient?

Reply:
Ultimately, my goal is that there is a single TC persistent (unstashable) header that represents the authenticated user so that the security plugin can reliably get information about the authenticated user from the TC and does not need to be concerned if the TC has been stashed.

The difference between User and UserSubject is that UserSubject has a `runAs` method that can be used to switch into the authenticated user context. For instance, its possible to write an API handler that switches to the plugin context and then back to the user context. 

The transient user header would be used to hold the value of the "user" to authorize actions with. In most circumstances, this would be the authenticated user, but when a plugin uses their pluginSubject and calls `runAs` then it will hold the identity associated with the plugin while the OPENDISTRO_SECURITY_AUTHENTICATED_USER header would hold information about the authenticated user.